File: Week-Wise-Assignments/Assignment 10/Systematic Solutions/C Programs/1_graph_adjacency_matrix.c
#include <stdio.h>
#include <stdlib.h>

#define MAX_VERTICES 100

typedef struct {
    int matrix[MAX_VERTICES][MAX_VERTICES];
    int vertices;
    int directed;
} Graph;

void initGraph(Graph* g, int vertices, int directed) {
    g->vertices = vertices;
    g->directed = directed;
    
    // Initialize matrix with 0s
    for (int i = 0; i < vertices; i++) {
        for (int j = 0; j < vertices; j++) {
            g->matrix[i][j] = 0;
        }
    }
    
    printf("Graph initialized with %d vertices (%s)\n", 
           vertices, directed ? "Directed" : "Undirected");
}

void addEdge(Graph* g, int src, int dest) {
    if (src >= 0 && src < g->vertices && dest >= 0 && dest < g->vertices) {
        g->matrix[src][dest] = 1;
        
        if (!g->directed) {
            g->matrix[dest][src] = 1;  // For undirected graph
        }
        
        printf("Edge added: %d -> %d\n", src, dest);
        if (!g->directed) {
            printf("Edge added: %d -> %d (undirected)\n", dest, src);
        }
    } else {
        printf("Invalid vertices: %d, %d\n", src, dest);
    }
}

void removeEdge(Graph* g, int src, int dest) {
    if (src >= 0 && src < g->vertices && dest >= 0 && dest < g->vertices) {
        g->matrix[src][dest] = 0;
        
        if (!g->directed) {
            g->matrix[dest][src] = 0;
        }
        
        printf("Edge removed: %d -> %d\n", src, dest);
    } else {
        printf("Invalid vertices: %d, %d\n", src, dest);
    }
}

int countVertices(Graph* g) {
    return g->vertices;
}

int countEdges(Graph* g) {
    int edges = 0;
    
    for (int i = 0; i < g->vertices; i++) {
        for (int j = 0; j < g->vertices; j++) {
            if (g->matrix[i][j] == 1) {
                edges++;
            }
        }
    }
    
    // For undirected graph, each edge is counted twice
    if (!g->directed) {
        edges /= 2;
    }
    
    return edges;
}

void findAdjacentVertices(Graph* g, int vertex) {
    if (vertex < 0 || vertex >= g->vertices) {
        printf("Invalid vertex: %d\n", vertex);
        return;
    }
    
    printf("Adjacent vertices of %d: ", vertex);
    int found = 0;
    
    for (int i = 0; i < g->vertices; i++) {
        if (g->matrix[vertex][i] == 1) {
            printf("%d ", i);
            found = 1;
        }
    }
    
    if (!found) {
        printf("None");
    }
    printf("\n");
}

int searchVertex(Graph* g, int vertex) {
    if (vertex >= 0 && vertex < g->vertices) {
        return 1;  // Vertex exists
    }
    return 0;  // Vertex doesn't exist
}

void displayGraph(Graph* g) {
    printf("\n=== Adjacency Matrix ===\n");
    printf("   ");
    for (int i = 0; i < g->vertices; i++) {
        printf("%2d ", i);
    }
    printf("\n");
    
    for (int i = 0; i < g->vertices; i++) {
        printf("%2d ", i);
        for (int j = 0; j < g->vertices; j++) {
            printf("%2d ", g->matrix[i][j]);
        }
        printf("\n");
    }
}

void graphAnalysis(Graph* g) {
    printf("\n=== Graph Analysis ===\n");
    printf("Number of vertices: %d\n", countVertices(g));
    printf("Number of edges: %d\n", countEdges(g));
    printf("Graph type: %s\n", g->directed ? "Directed" : "Undirected");
    
    // Calculate degree for each vertex
    printf("\nVertex degrees:\n");
    for (int i = 0; i < g->vertices; i++) {
        int inDegree = 0, outDegree = 0;
        
        // Calculate out-degree
        for (int j = 0; j < g->vertices; j++) {
            if (g->matrix[i][j] == 1) {
                outDegree++;
            }
        }
        
        // Calculate in-degree
        for (int j = 0; j < g->vertices; j++) {
            if (g->matrix[j][i] == 1) {
                inDegree++;
            }
        }
        
        if (g->directed) {
            printf("Vertex %d: In-degree = %d, Out-degree = %d\n", 
                   i, inDegree, outDegree);
        } else {
            printf("Vertex %d: Degree = %d\n", i, outDegree);
        }
    }
    
    // Memory usage
    int memoryUsage = g->vertices * g->vertices * sizeof(int);
    printf("\nMemory usage: %d bytes\n", memoryUsage);
    printf("Space complexity: O(V²) where V = %d\n", g->vertices);
}

void createGraphInteractive(Graph* g) {
    int vertices, directed, edges, src, dest;
    
    printf("\n=== Creating Graph ===\n");
    printf("Enter number of vertices: ");
    scanf("%d", &vertices);
    
    if (vertices <= 0 || vertices > MAX_VERTICES) {
        printf("Invalid number of vertices!\n");
        return;
    }
    
    printf("Is the graph directed? (1 for Yes, 0 for No): ");
    scanf("%d", &directed);
    
    initGraph(g, vertices, directed);
    
    printf("Enter number of edges: ");
    scanf("%d", &edges);
    
    printf("Enter edges (source destination):\n");
    for (int i = 0; i < edges; i++) {
        printf("Edge %d: ", i + 1);
        scanf("%d %d", &src, &dest);
        addEdge(g, src, dest);
    }
    
    printf("Graph created successfully!\n");
}

void dfsUtil(Graph* g, int vertex, int visited[]) {
    visited[vertex] = 1;
    printf("%d ", vertex);
    
    for (int i = 0; i < g->vertices; i++) {
        if (g->matrix[vertex][i] == 1 && !visited[i]) {
            dfsUtil(g, i, visited);
        }
    }
}

void dfs(Graph* g, int startVertex) {
    if (startVertex < 0 || startVertex >= g->vertices) {
        printf("Invalid start vertex!\n");
        return;
    }
    
    int visited[MAX_VERTICES] = {0};
    
    printf("DFS traversal starting from vertex %d: ", startVertex);
    dfsUtil(g, startVertex, visited);
    printf("\n");
}

void bfs(Graph* g, int startVertex) {
    if (startVertex < 0 || startVertex >= g->vertices) {
        printf("Invalid start vertex!\n");
        return;
    }
    
    int visited[MAX_VERTICES] = {0};
    int queue[MAX_VERTICES];
    int front = 0, rear = 0;
    
    visited[startVertex] = 1;
    queue[rear++] = startVertex;
    
    printf("BFS traversal starting from vertex %d: ", startVertex);
    
    while (front < rear) {
        int vertex = queue[front++];
        printf("%d ", vertex);
        
        for (int i = 0; i < g->vertices; i++) {
            if (g->matrix[vertex][i] == 1 && !visited[i]) {
                visited[i] = 1;
                queue[rear++] = i;
            }
        }
    }
    printf("\n");
}

int main() {
    Graph g;
    int choice, vertex, src, dest, result;
    
    printf("=== Graph Implementation (Adjacency Matrix) ===\n");
    printf("Adjacency Matrix: 2D array representation\n");
    printf("Space Complexity: O(V²), Time Complexity: O(1) for edge operations\n\n");
    
    // Initialize with default values
    initGraph(&g, 0, 0);
    
    while (1) {
        printf("\n--- Menu ---\n");
        printf("1. Create Graph\n");
        printf("2. Add Edge\n");
        printf("3. Remove Edge\n");
        printf("4. Count Vertices\n");
        printf("5. Count Edges\n");
        printf("6. Find Adjacent Vertices\n");
        printf("7. Search Vertex\n");
        printf("8. Display Graph\n");
        printf("9. Graph Analysis\n");
        printf("10. DFS Traversal\n");
        printf("11. BFS Traversal\n");
        printf("12. Exit\n");
        printf("Enter your choice: ");
        scanf("%d", &choice);
        
        switch (choice) {
            case 1:
                createGraphInteractive(&g);
                break;
                
            case 2:
                printf("Enter source and destination vertices: ");
                scanf("%d %d", &src, &dest);
                addEdge(&g, src, dest);
                break;
                
            case 3:
                printf("Enter source and destination vertices: ");
                scanf("%d %d", &src, &dest);
                removeEdge(&g, src, dest);
                break;
                
            case 4:
                result = countVertices(&g);
                printf("Number of vertices: %d\n", result);
                break;
                
            case 5:
                result = countEdges(&g);
                printf("Number of edges: %d\n", result);
                break;
                
            case 6:
                printf("Enter vertex: ");
                scanf("%d", &vertex);
                findAdjacentVertices(&g, vertex);
                break;
                
            case 7:
                printf("Enter vertex to search: ");
                scanf("%d", &vertex);
                result = searchVertex(&g, vertex);
                printf("Vertex %d %s in the graph\n", 
                       vertex, result ? "exists" : "does not exist");
                break;
                
            case 8:
                displayGraph(&g);
                break;
                
            case 9:
                graphAnalysis(&g);
                break;
                
            case 10:
                printf("Enter starting vertex for DFS: ");
                scanf("%d", &vertex);
                dfs(&g, vertex);
                break;
                
            case 11:
                printf("Enter starting vertex for BFS: ");
                scanf("%d", &vertex);
                bfs(&g, vertex);
                break;
                
            case 12:
                printf("Exiting program...\n");
                exit(0);
                
            default:
                printf("Invalid choice! Please try again.\n");
        }
    }
    
    return 0;
}

File: Week-Wise-Assignments/Assignment 10/Systematic Solutions/Python Programs/1_graph_adjacency_matrix.py
from collections import deque

class GraphMatrix:
    def __init__(self, vertices=0, directed=False):
        self.vertices = vertices
        self.directed = directed
        self.matrix = [[0 for _ in range(vertices)] for _ in range(vertices)]
    
    def add_edge(self, src, dest):
        """Add an edge between source and destination vertices"""
        if 0 <= src < self.vertices and 0 <= dest < self.vertices:
            self.matrix[src][dest] = 1
            if not self.directed:
                self.matrix[dest][src] = 1
            print(f"Edge added: {src} -> {dest}")
            if not self.directed:
                print(f"Edge added: {dest} -> {src} (undirected)")
        else:
            print(f"Invalid vertices: {src}, {dest}")
    
    def remove_edge(self, src, dest):
        """Remove an edge between source and destination vertices"""
        if 0 <= src < self.vertices and 0 <= dest < self.vertices:
            self.matrix[src][dest] = 0
            if not self.directed:
                self.matrix[dest][src] = 0
            print(f"Edge removed: {src} -> {dest}")
        else:
            print(f"Invalid vertices: {src}, {dest}")
    
    def count_vertices(self):
        """Count number of vertices in the graph"""
        return self.vertices
    
    def count_edges(self):
        """Count number of edges in the graph"""
        edges = 0
        for i in range(self.vertices):
            for j in range(self.vertices):
                if self.matrix[i][j] == 1:
                    edges += 1
        
        # For undirected graph, each edge is counted twice
        if not self.directed:
            edges //= 2
        
        return edges
    
    def find_adjacent_vertices(self, vertex):
        """Find all adjacent vertices of a given vertex"""
        if 0 <= vertex < self.vertices:
            adjacent = []
            for i in range(self.vertices):
                if self.matrix[vertex][i] == 1:
                    adjacent.append(i)
            return adjacent
        else:
            print(f"Invalid vertex: {vertex}")
            return []
    
    def search_vertex(self, vertex):
        """Search if a vertex exists in the graph"""
        return 0 <= vertex < self.vertices
    
    def display_graph(self):
        """Display the adjacency matrix"""
        print("\n=== Adjacency Matrix ===")
        print("   ", end="")
        for i in range(self.vertices):
            print(f"{i:2d} ", end="")
        print()
        
        for i in range(self.vertices):
            print(f"{i:2d} ", end="")
            for j in range(self.vertices):
                print(f"{self.matrix[i][j]:2d} ", end="")
            print()
    
    def get_degree(self, vertex):
        """Get degree of a vertex"""
        if 0 <= vertex < self.vertices:
            if self.directed:
                in_degree = sum(self.matrix[i][vertex] for i in range(self.vertices))
                out_degree = sum(self.matrix[vertex][i] for i in range(self.vertices))
                return in_degree, out_degree
            else:
                degree = sum(self.matrix[vertex][i] for i in range(self.vertices))
                return degree
        return None
    
    def graph_analysis(self):
        """Comprehensive graph analysis"""
        print("\n=== Graph Analysis ===")
        print(f"Number of vertices: {self.count_vertices()}")
        print(f"Number of edges: {self.count_edges()}")
        print(f"Graph type: {'Directed' if self.directed else 'Undirected'}")
        
        print(f"\nVertex degrees:")
        for i in range(self.vertices):
            if self.directed:
                in_deg, out_deg = self.get_degree(i)
                print(f"Vertex {i}: In-degree = {in_deg}, Out-degree = {out_deg}")
            else:
                degree = self.get_degree(i)
                print(f"Vertex {i}: Degree = {degree}")
        
        # Memory usage
        memory_usage = self.vertices * self.vertices * 4  # Assuming 4 bytes per integer
        print(f"\nMemory usage: {memory_usage} bytes")
        print(f"Space complexity: O(V²) where V = {self.vertices}")
    
    def dfs(self, start_vertex):
        """Depth-First Search traversal"""
        if not (0 <= start_vertex < self.vertices):
            print("Invalid start vertex!")
            return []
        
        visited = [False] * self.vertices
        result = []
        
        def dfs_util(vertex):
            visited[vertex] = True
            result.append(vertex)
            
            for i in range(self.vertices):
                if self.matrix[vertex][i] == 1 and not visited[i]:
                    dfs_util(i)
        
        dfs_util(start_vertex)
        return result
    
    def bfs(self, start_vertex):
        """Breadth-First Search traversal"""
        if not (0 <= start_vertex < self.vertices):
            print("Invalid start vertex!")
            return []
        
        visited = [False] * self.vertices
        queue = deque([start_vertex])
        visited[start_vertex] = True
        result = []
        
        while queue:
            vertex = queue.popleft()
            result.append(vertex)
            
            for i in range(self.vertices):
                if self.matrix[vertex][i] == 1 and not visited[i]:
                    visited[i] = True
                    queue.append(i)
        
        return result
    
    def create_graph_interactive(self):
        """Create graph interactively"""
        print("\n=== Creating Graph ===")
        try:
            vertices = int(input("Enter number of vertices: "))
            if vertices <= 0:
                print("Invalid number of vertices!")
                return
            
            directed = input("Is the graph directed? (y/n): ").lower() == 'y'
            
            self.vertices = vertices
            self.directed = directed
            self.matrix = [[0 for _ in range(vertices)] for _ in range(vertices)]
            
            edges = int(input("Enter number of edges: "))
            
            print("Enter edges (source destination):")
            for i in range(edges):
                src, dest = map(int, input(f"Edge {i + 1}: ").split())
                self.add_edge(src, dest)
            
            print("Graph created successfully!")
            
        except ValueError:
            print("Invalid input! Please enter integers only.")

def main():
    graph = GraphMatrix()
    
    print("=== Graph Implementation (Adjacency Matrix) ===")
    print("Adjacency Matrix: 2D array representation")
    print("Space Complexity: O(V²), Time Complexity: O(1) for edge operations\n")
    
    while True:
        print("\n--- Menu ---")
        print("1. Create Graph")
        print("2. Add Edge")
        print("3. Remove Edge")
        print("4. Count Vertices")
        print("5. Count Edges")
        print("6. Find Adjacent Vertices")
        print("7. Search Vertex")
        print("8. Display Graph")
        print("9. Graph Analysis")
        print("10. DFS Traversal")
        print("11. BFS Traversal")
        print("12. Exit")
        
        try:
            choice = int(input("Enter your choice: "))
            
            if choice == 1:
                graph.create_graph_interactive()
            
            elif choice == 2:
                src, dest = map(int, input("Enter source and destination vertices: ").split())
                graph.add_edge(src, dest)
            
            elif choice == 3:
                src, dest = map(int, input("Enter source and destination vertices: ").split())
                graph.remove_edge(src, dest)
            
            elif choice == 4:
                count = graph.count_vertices()
                print(f"Number of vertices: {count}")
            
            elif choice == 5:
                count = graph.count_edges()
                print(f"Number of edges: {count}")
            
            elif choice == 6:
                vertex = int(input("Enter vertex: "))
                adjacent = graph.find_adjacent_vertices(vertex)
                print(f"Adjacent vertices of {vertex}: {adjacent}")
            
            elif choice == 7:
                vertex = int(input("Enter vertex to search: "))
                exists = graph.search_vertex(vertex)
                print(f"Vertex {vertex} {'exists' if exists else 'does not exist'} in the graph")
            
            elif choice == 8:
                graph.display_graph()
            
            elif choice == 9:
                graph.graph_analysis()
            
            elif choice == 10:
                start = int(input("Enter starting vertex for DFS: "))
                result = graph.dfs(start)
                print(f"DFS traversal starting from vertex {start}: {result}")
            
            elif choice == 11:
                start = int(input("Enter starting vertex for BFS: "))
                result = graph.bfs(start)
                print(f"BFS traversal starting from vertex {start}: {result}")
            
            elif choice == 12:
                print("Exiting program...")
                break
            
            else:
                print("Invalid choice! Please try again.")
        
        except ValueError:
            print("Invalid input! Please enter a valid number.")
        except KeyboardInterrupt:
            print("\nProgram interrupted. Exiting...")
            break

if __name__ == "__main__":
    main()